goldmane/proto/api.proto

```diff
@@ -47,15 +34,111 @@ message FlowRequest {
   // PageSize configures the maximum number of results to return as part of this query.
   int64 page_size = 4;
 
-  // SortBy configures how to sort the results of this query. By default flows are time-sorted.
-  SortBy sort_by = 5;
+  // SortBy configures how to sort the results of this query. By default flows are sorted by start time.
+  // The returned list is sorted by each sort option, in order, using the next sort option in the list as a tie-breaker.
+  // Note: At the moment, only a single sort option is supported.
+  repeated SortOption sort_by = 5;
+
+  // Filter allows specification of one or more criteria on which to filter the returned Flows.
+  Filter filter = 6;
+
+  // AggregationInterval is the width of the time window in seconds across which to aggregate when generating
+  // Flows to return. This must be a multiple of 15.
+  int64 aggregation_interval = 7;
+}
+
+// StreamRequest defines a message to request a stream of aggregated Flows.
+message StreamRequest {
+  // StartTimeGt specifies the beginning of a time window from which to stream Flows. Flows
+  // will be streamed only if their start time occurs after the requested time.
+  int64 start_time_gt = 1;
+
+  // Filter allows specification of one or more criteria on which to filter the returned Flows.
+  Filter filter = 2;
+
+  // AggregationInterval defines both the frequency of streamed updates for each Flow, and the amount of time that FlowResult covers.
+  // It must always be 15s.
+  //
+  // Every AggregationInterval the server must send a FlowResult containing the aggregated data for that Flow from a
+  // time interval of width AggregationInterval.
+  //
+  // For a Flow that has continuous traffic, the server should send updates covering the range
+  // [now-2*AggregationInterval, now-AggregationInterval] so that the data is reasonably likely to be complete.
+  int64 aggregation_interval = 3;
+}
+
+// FlowResult wraps a Flow object with additional metadata.
+message FlowResult {
+  // ID is an opaque integer value ID that can be used to identify a Flow, and is 1:1 with the FlowKey.
```

Comment:
What were the issues with adding this ID to the Flow and saying 0 means "not set"?

Reply:
A Flow is used on multiple APIs, including those that don't have a concept of an ID - e.g., the push API to Linseed and the collection API used by Felix.

This ID is only a concept on the aggregated List/Stream APIs, and I think it makes sense that those APIs may want a place to include additional metadata in the future anyway.

Reply:
I had some other thoughts on this, but realize it probably doesn't matter for this so I'm fine with it as is.